### PR TITLE
WIP fix for shell script executable bit and scaffolding for template tests

### PR DESCRIPTION
**Pull request by baronfel**

### CODE REVIEW

**Comment by baronfel** [comment]
GAH! We can't actually set the executable bit because of a few outstanding templating/coreclr bugs: https://github.com/dotnet/templating/issues/1547

However we should still add this because a) it'll be fixed at _some_ point, and b) even if it fails it leaves a helpful error message:
```
The files /private/var/folders/80/y0q8bqgx29bd8g3h0tlt00k00000gn/T/mrisxb4t.jrn/chmod, /private/var/folders/80/y0q8bqgx29bd8g3h0tlt00k00000gn/T/mrisxb4t.jrn/+x, and /private/var/folders/80/y0q8bqgx29bd8g3h0tlt00k00000gn/T/mrisxb4t.jrn/*.sh do not exist.
Unable to apply permissions +x to "*.sh".
Post action failed.
Description: Make scripts executable
Manual instructions: Run 'chmod +x *.sh'
```

**Comment by matthid** [comment]
@baronfel I don't think we can expect them to change the behavior again, the flag just was used incorrectly all the years - I agree that using it like this is not portable at all. Therefore we should probably change to `/bin/sh -c` or `/usr/bin/env sh -c `

**Comment by baronfel** [comment]
that's fair 👍

**Comment by baronfel** [comment]
hard to test the `run an executable` action in an integration test though because the promtps require user input: 
```
Template is configured to run the following action:
Description: Make scripts executable
Manual instructions: Run 'chmod +x *.sh'
Actual command: /bin/sh -c "echo pwd && chmod +x *.sh"
Do you want to run this action (Y|N)?
```

**Comment by baronfel** [comment]
Ok, I've added a test for both the tool and project variants. It would be easy to write more. The hard part will be interacting with the tests because now we have to answer 'yes' for each test of the template.  For this reason I've marked them _pending_ but I think we should use them when working on the template locally for sure, at least until the input question is solved.

**Comment by matthid** [comment]
We can either try to set all options beforehand via command line or just redirect and write standard input

**Comment by matthid** [comment]
Apparently there is `--allow-scripts yes`, see https://github.com/dotnet/templating/issues/765#issuecomment-299677044 
Seems to be undocumented for now as I cannot see it in `--help`

File: src/test/Fake.Dotnet.Cli.IntegrationTests/TemplateTests.fs [hunk 1]
```diff
@@ -1,11 +1,75 @@
-﻿module Fake.Dotnet.Cli.IntegrationTests.TemplateTests
+﻿module Fake.DotNet.Cli.IntegrationTests.TemplateTests
```
Comment (matthid) [inline src/test/Fake.Dotnet.Cli.IntegrationTests/TemplateTests.fs]: Please also rename the folder in git, otherwise there will be problems (which is a bit unfortunate to do on windows as you have to temporarily rename to something else)

**Comment by baronfel** [comment]
Ok, I did it via redirecting stdin

File: src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+let invokeScript dir scriptName =
+    let fullScriptPath = Path.Combine(dir, scriptName)
+    Process.execWithResult 
+        (fun x -> 
+            x.WithWorkingDirectory(dir)
+             .WithFileName(fullScriptPath)
+             .WithArguments "--help" ) (System.TimeSpan.FromSeconds 60.)
+    |> shouldSucceed "should invoke the script file"
```
Comment (matthid) [inline src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: I'd prefer a test which actually checks if `./fake build` works (ie all dependencies still work)
Also can we add a failing test (ie that the scripts correctly return nonzero exit-codes on failure). For example by executing `./fake.sh run nonexisting.fsx`.

File: src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            p.WithWorkingDirectory(rootDir)
+             .WithFileName(dotnet)
+             .WithArguments(sprintf "new %s --bootstrap %s" templateName (string kind))
+             .WithRedirectStandardInput(true)
+             .WithRedirectStandardOutput(true)
+             .WithRedirectStandardError(true)
+        )
+    
+    proc.Start () |> ignore
+    do! Async.Sleep 2000
```
Comment (matthid) [inline src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: Do we really need that?

**Comment by matthid** [comment]
Thanks a lot for taking care of this! Just two more things I noticed.

File: src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs (hunk 3 above)
Comment (baronfel) [reply src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: Yeah, I was guessing at how long to wait after invoking the template gen, so that the prompt would be active at the point that I sent the Y character.

File: src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs (hunk 2 above)
Comment (baronfel) [reply src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: Can do

File: src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs (hunk 3 above)
Comment (matthid) [reply src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: Did you try `--allow-scripts yes` (not the most important - but I guess it would not introduce a potential race condition)

Comment (baronfel) [reply src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: No, I'll give that a try when I add the new tests

**Comment by baronfel** [comment]
Is there anything else here I missed that you'd like me to cover?

**Comment by matthid** [comment]
Only the one comment, but otherwise looks good

**Comment by baronfel** [comment]
I'm sorry, which comment do you mean? I thought I just needed to look into `--allow-scripts` (which we're now using successfully) and adding some more tests around the positive/negative template-script uses (which I added).

File: src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                runTemplate tempDir Project
+                invokeScript tempDir scriptFile "build -t All" |> shouldSucceed "should build successfully"
+            }
+
+            yield test "fails to build a target that doesn't exist" {
+                let tempDir = Path.Combine(Path.GetTempPath (), Path.GetRandomFileName())
+                Directory.ensure tempDir
+                runTemplate tempDir Project
+                let result = invokeScript tempDir scriptFile "build -t Nonexistent" 
+                Expect.isFalse (result.OK && missingTarget "Nonexistent" result) "Should fail to build a target that doesn't exist"
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/test/Fake.DotNet.Cli.IntegrationTests/TemplateTests.fs]: ah you are of course correct. I updated the test checks to be more clear about the two conditions.

**Comment by matthid** [comment]
Oh we do not actually run the new tests in the build process, do we?

**Comment by baronfel** [comment]
ah yeah. I'm looking at how to place them post-dotnet-pack right now.

**Comment by matthid** [comment]
You can also integrate it into the `DotNetCoreIntegrationTests` just call dotnet run like the others

**Comment by baronfel** [comment]
Yeah but I'm relying on the pack task having packaged the nuget package, so I can just install it. That seems harder to orchestrate than just running the template tests sometime after packaging. What do you think?

**Comment by matthid** [comment]
Ah yeah whatever you think is easier :)

**Comment by baronfel** [comment]
Ok, I think I got them. The template tests are a little fragile because they point to the nuget directory precisely, but they do actually get invoked now.

**Comment by matthid** [comment]
Thanks a lot for taking care of this thing!

**Comment by rmunn** [comment]
https://github.com/dotnet/templating/issues/1547 has been fixed by https://github.com/dotnet/templating/pull/1555, so I suspect it'll be in the next .Net Core bugfix release. But since version 2.1.300 was released with the bug in it, it'll be a while before it's safe to use the `chmod` post-action (the one with GUID cb9a6cf3-4f5c-4860-b9d2-03a574959774) and trust that it'll work on *everyone*'s system.
